File: frontend/src/types/branchingDrama.ts
import type { FusionToArcOperatorMode } from '@/types/fusionStory'

export type BranchingDramaStatus =
  | 'generating_outlines'
  | 'generating_images'
  | 'ready'
  | 'failed'

/**
 * Where the player stands inside the drama (BD2). Closed vocabulary —
 * the backend plans and narrates the whole tree against this value:
 * `central` (主演), `present` (在場配角), `absent` (旁觀者).
 */
export type DramaOperatorPosition = 'central' | 'present' | 'absent'

export const DRAMA_OPERATOR_POSITIONS: DramaOperatorPosition[] = [
  'central',
  'present',
  'absent',
]

/** What the create form starts on, matching the backend default. */
export const DEFAULT_DRAMA_OPERATOR_POSITION: DramaOperatorPosition = 'central'

/** Mirrors the entity ceiling (`OPERATOR_NOTE_MAX_CHARS`). */
export const DRAMA_OPERATOR_NOTE_MAX_CHARS = 120

/**
 * The look every scene image of one drama is drawn in (BD10). Closed
 * vocabulary, shared with the per-character `visual_generation_style` —
 * except that the character-level slot has a fourth `''` ("inherit") state
 * and this one does not: a drama is one production and states its own art
 * direction outright.
 */
export type DramaVisualStyle = 'anime' | 'realistic'

export const DRAMA_VISUAL_STYLES: DramaVisualStyle[] = ['anime', 'realistic']

/** Matches `VISUAL_GENERATION_STYLE_DEFAULT` on the backend. */
export const DEFAULT_DRAMA_VISUAL_STYLE: DramaVisualStyle = 'anime'

/**
 * Mirrors the *create-path* ceiling (`MAX_TOTAL_SEGMENTS`, BD4). Node count
 * grows as `3^N`, so this bounds worst-case generation fan-out — keep in
 * sync with `BranchingDrama.create_pending` and the DTO's `le=` in
 * `application/dto/branching_drama.py`.
 *
 * It bounds the create form only. Dramas made under an older, higher cap
 * still list and play, so nothing that *renders* a drama may clamp to this
 * number — the entity constructor deliberately does not either (FX3).
 */
export const MAX_DRAMA_TOTAL_SEGMENTS = 12

/** The other half of the same DTO bound (`ge=2`): a drama needs a branch. */
export const MIN_DRAMA_TOTAL_SEGMENTS = 2

/**
 * The three-way branch every non-root node offers (mirrors `VALID_TONES` /
 * `TONE_DARK` / `TONE_SUNNY` / `TONE_NEUTRAL` in `branching_drama.py`).
 * Closed vocabulary — the wire value is this English enum, and it must
 * never reach the player un-translated (BD11): route it through
 * `dramaToneLabel` in `utils/dramaTone.ts` instead of rendering it raw.
 */
export type DramaTone = 'dark' | 'sunny' | 'neutral'

export const DRAMA_TONES: DramaTone[] = ['dark', 'sunny', 'neutral']

export interface DramaNode {
  id: string
  drama_id: string
  parent_node_id: string | null
  depth: number
  tone: string | null
  title: string
  summary: string
  appearing_character_ids: string[]
  image_path: string | null
}

/**
 * One picture the player walked past (BD9). Safe to show whole — it is a
 * scene they already read.
 */
export interface CollectedScene {
  node_id: string
  title: string
  depth: number
  tone: string | null
  image_path: string
}

/**
 * 劇場圖集 — what this player collected out of one drama's painted tree.
 *
 * The un-walked pictures arrive as `locked_count` and nothing else: the
 * backend never sends their titles, summaries or paths, so a silhouette tile
 * has nothing to spoil even if a future view tried to render more. Do not
 * add fields here expecting the server to fill them — the omission is the
 * feature.
 *
 * `total_with_images` counts the nodes of the *whole* tree that lazy
 * generation has already painted — walked or not — so it runs *ahead* of the
 * player: the prefetch paints branches nobody ever enters. That is exactly
 * why it is no longer a progress denominator (D8.2): the 「已收集 X/Y」 ratio
 * it produced was both low and flat — pinned near 1/3 only at
 * `KOKORO_DRAMA_IMAGE_PREFETCH_DEPTH=1`, down in the 1/9–15% band at the
 * default depth of 2 — and never equal to what the player walked. Completion is
 * measured against the whole tree instead (`utils/dramaProgress`). The field
 * stays on the wire untouched; nothing on this page divides by it any more —
 * and it is not a count of the pictures this player collected either.
 */
export interface DramaSceneGallery {
  collected: CollectedScene[]
  locked_count: number
  total_with_images: number
}

export interface Exchange {
  player_input: string
  response: string
}

export interface DramaSessionTurn {
  node_id: string
  narration: string
  player_input: string
  chosen_tone: string | null
  exchanges: Exchange[]
}

export interface DramaSession {
  id: string
  drama_id: string
  current_node_id: string
  status: 'playing' | 'ended'
  turns: DramaSessionTurn[]
  created_at: string
  updated_at: string
}

export interface BranchingDrama {
  id: string
  character_ids: string[]
  prompt: string
  title: string
  total_segments: number
  status: BranchingDramaStatus
  error_message: string | null
  /**
   * Machine-readable failure reason for a `failed` job (plan U4b), from
   * an open set the backend forwards verbatim. `insufficient_credits`
   * is the only value the SPA acts on; anything else (or `null`, for an
   * ordinary crash) falls through to the generic failure copy.
   */
  error_code?: string | null
  operator_position: DramaOperatorPosition
  operator_note: string | null
  /**
   * The look this drama's scene images are drawn in (BD10). `null` only for
   * a drama created before the slot existed — those still style themselves
   * off their first character, so the detail view must render nothing
   * rather than claim a look they do not have.
   */
  visual_style: DramaVisualStyle | null
  expected_node_count: number
  /**
   * Node count pre-generated synchronously at creation (root + prefetch
   * layers) — the correct denominator for the early `generating_outlines`
   * progress bar. `expected_node_count` above is the full-tree total
   * (only ever fully realized after lazy generation catches up while the
   * player advances), so dividing by it there pins the bar near 0%.
   */
  initial_node_target: number
  generated_node_count: number
  first_scene_image_path: string | null
  /**
   * Root node id, so the detail view's title image can name the node it
   * wants redrawn (BD6). Present even when `first_scene_image_path` is
   * `null` — a node with no picture is exactly what the redraw repairs.
   */
  first_scene_node_id: string | null
  warning: string | null
  created_at: string
  updated_at: string
}

export interface BranchingDramaSummary {
  id: string
  character_ids: string[]
  title: string
  total_segments: number
  status: BranchingDramaStatus
  error_message: string | null
  error_code?: string | null
  created_at: string
  updated_at: string
}

/**
 * The fixed price this client had on screen when the player pressed (BD3).
 * The server binds the charge to it and answers `409 price_changed` if the
 * published price moved, rather than silently charging a number nobody saw.
 * Absent whenever nothing is quotable — self-host, a token-billed tier, a
 * price list that never loaded — in which case the server falls back to its
 * own cache exactly as before.
 */
export interface QuotedActionPrice {
  quoted_price_cr?: number
}

export interface CreateBranchingDramaPayload extends QuotedActionPrice {
  character_ids: string[]
  prompt: string
  total_segments: number
  /**
   * Optional on the wire: the backend defaults it to `central`, so an
   * entry point that has no opinion (the fusion 「換個玩法」 handoff) can
   * simply leave it out and get the framing dramas always had.
   */
  operator_position?: DramaOperatorPosition
  operator_note?: string | null
  /**
   * Optional on the wire for the same reason `operator_position` is: an
   * entry point with no opinion (the fusion 「換個玩法」 handoff) leaves it
   * out, and the backend resolves the first character's style *and stores
   * it*. Omitting it never means "anime" — it means "you decide, once".
   */
  visual_style?: DramaVisualStyle
}

/**
 * What `interact` posts. Named rather than inlined at the call site because
 * an object literal handed straight to the generic `withQuotedPrice` is
 * excess-property-checked against its `QuotedActionPrice` constraint, which
 * rejects `player_input` outright.
 */
export interface InteractSessionPayload extends QuotedActionPrice {
  player_input: string
}

export interface InteractSessionResponse {
  session: DramaSession
  response: string
  advance_hint: string | null
}

export interface AdvanceSessionResponse {
  session: DramaSession
  current_node: DramaNode
  is_ending: boolean
}

/**
 * BD7 — 把走過的這條路寫成劇本. Both fields optional in both directions:
 * an omitted `operator_mode` is a player who did not touch the chips, and
 * the backend answers that with the framing the drama was actually played
 * under (see `DRAMA_POSITION_TO_ARC_MODE`) rather than a blanket default.
 */
export interface DramaToArcDraftPayload {
  instruction?: string
  operator_mode?: FusionToArcOperatorMode
}

/**
 * The conversion mode a drama's own player position implies (BD7).
 *
 * A pre-fill, never a lock — the exit hub shows all three chips, and a
 * player who watched a drama may still ask to be written into the arc it
 * becomes. Kept beside the position vocabulary it reads so the two closed
 * sets cannot drift apart silently; the backend holds the same table
 * (`contracts/drama_to_arc.operator_mode_for_drama_position`) because the
 * server must answer a client that sends nothing at all.
 */
export const DRAMA_POSITION_TO_ARC_MODE: Record<
  DramaOperatorPosition,
  FusionToArcOperatorMode
> = {
  central: 'write_in',
  present: 'observer',
  absent: 'unchanged',
}

File: src/kokoro_link/application/dto/branching_drama.py
"""DTOs for the branching-drama REST API."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field, field_validator

from kokoro_link.application.services.branching_drama_gallery import (
    CollectedScene,
    DramaSceneGallery,
)
from kokoro_link.contracts.fusion_to_arc import (
    VALID_FUSION_OPERATOR_MODES,
    normalise_fusion_operator_mode,
)
from kokoro_link.domain.entities.branching_drama import (
    BranchingDrama,
    DEFAULT_OPERATOR_POSITION,
    DEFAULT_TOTAL_SEGMENTS,
    DramaNode,
    DramaSession,
    DramaSessionTurn,
    Exchange,
    MAX_TOTAL_SEGMENTS,
    MIN_TOTAL_SEGMENTS,
    OPERATOR_NOTE_MAX_CHARS,
    SEGMENTS_WARNING_THRESHOLD,
    _MAX_CHARACTERS,
    _MIN_CHARACTERS,
    normalise_drama_visual_style,
    normalise_operator_position,
)


# ── requests ──────────────────────────────────────────────────────────


class CreateBranchingDramaRequest(BaseModel):
    character_ids: list[str] = Field(
        ..., min_length=_MIN_CHARACTERS, max_length=_MAX_CHARACTERS,
    )
    prompt: str = Field(..., min_length=1, max_length=2000)
    total_segments: int = Field(
        default=DEFAULT_TOTAL_SEGMENTS,
        ge=MIN_TOTAL_SEGMENTS,
        le=MAX_TOTAL_SEGMENTS,
    )
    """Bounded here *and* in ``create_pending`` — the create path is where
    the BD4 ceiling is enforced, because the entity constructor doubles as
    the row-reading path and must stay tolerant of dramas created under an
    older, higher cap (FX3)."""
    operator_position: str = Field(default=DEFAULT_OPERATOR_POSITION)
    """Where the player stands in the drama (BD2). Defaulted rather than
    required so a client that predates the slot creates exactly the drama
    it always created."""
    operator_note: str | None = Field(
        default=None, max_length=OPERATOR_NOTE_MAX_CHARS,
    )
    visual_style: str | None = Field(default=None, max_length=32)
    """Which look this drama's scene images are drawn in (BD10).

    Optional rather than required, and ``None`` is *not* "anime": it means
    the client did not state a look — the fusion 「換個玩法」 handoff, a
    pre-BD10 client — and the service answers it by resolving the first
    character's style at create time and **storing that**. So the implicit
    answer becomes an explicit one on the way in, and every drama made
    from here on has a look written on it."""
    quoted_price_cr: float | None = None
    """The fixed price this client had on screen when the player pressed
    create (BD3). The charge binds to it, so a back-office edit mid-session
    refuses with ``409 price_changed`` instead of billing a number nobody was
    shown. ``None`` = nothing quotable (self-host, a token-billed tier, a
    price list that never loaded) and the server falls back to its own cache."""

    @field_validator("operator_position")
    @classmethod
    def _check_position(cls, value: str) -> str:
        """Reject an off-vocabulary position at the edge (422).

        Reuses the domain normaliser so casing is canonicalised and the
        legal set cannot drift from the entity's.
        """
        return normalise_operator_position(value)

    @field_validator("visual_style")
    @classmethod
    def _check_visual_style(cls, value: str | None) -> str | None:
        """Reject an off-vocabulary look at the edge (422).

        Same domain normaliser the entity uses, so casing is canonicalised
        and blank collapses to ``None`` (= "not stated") rather than to a
        style nobody picked.
        """
        return normalise_drama_visual_style(value)


class InteractSessionRequest(BaseModel):
    player_input: str = Field(..., min_length=1, max_length=4000)
    quoted_price_cr: float | None = None
    """Same binding as create — the price on the player's screen (BD3)."""


class AdvanceSessionRequest(BaseModel):
    """Optional body on ``/advance`` — it exists only to carry the quote.

    Advance took no body before BD3 and still accepts none: a client that
    posts nothing advances exactly as it always did, and the server quotes
    from its own price cache. Hosted clients post the number they displayed
    so the charge binds to it.
    """

    quoted_price_cr: float | None = None


class StartSessionRequest(BaseModel):
    """Optional body on ``/sessions`` — it exists only to carry the quote.

    Starting a playthrough took no body before FX1 and still accepts none.
    The opening is charged as one ``branching_drama_advance`` (it *is* the
    zeroth press), so a hosted client posts the advance price it displayed
    and the charge binds to that number; anything else posts nothing and the
    server quotes from its own cache.
    """

    quoted_price_cr: float | None = None


class RegenerateSceneImageRequest(BaseModel):
    """Optional body on the manual scene redraw — it carries the quote only.

    Same shape and same reason as :class:`AdvanceSessionRequest`: a client
    with nothing quotable (self-host, a token-billed tier, a price list that
    never loaded) posts nothing and the server falls back to its own cache.
    """

    quoted_price_cr: float | None = None


class DramaToArcDraftRequest(BaseModel):
    """Optional body on 「把這條路寫成劇本」 (BD7).

    Both fields are optional in both directions. An omitted body is a
    player who pressed the button without touching the mode chips, and
    the service answers that with the framing the drama was actually
    played under (``operator_position``) rather than a blanket default —
    unlike the fusion path, this source knows where the player stood.
    An off-vocabulary mode is a 422, never a best-guess branch.
    """

    instruction: str | None = Field(default=None, max_length=2000)
    operator_mode: str | None = Field(default=None, max_length=32)

    @field_validator("operator_mode")
    @classmethod
    def _check_operator_mode(cls, value: str | None) -> str | None:
        # Blank collapses to "unanswered" rather than to the fusion
        # protocol default: on this path an unanswered mode has a better
        # answer waiting (the drama's own position), and letting an empty
        # string mean ``unchanged`` would silently overrule it.
        if value is None or not value.strip():
            return None
        try:
            return normalise_fusion_operator_mode(value)
        except ValueError as exc:
            raise ValueError(
                "operator_mode must be one of "
                f"{sorted(VALID_FUSION_OPERATOR_MODES)}",
            ) from exc


# ── responses ─────────────────────────────────────────────────────────


class DramaNodeResponse(BaseModel):
    id: str
    drama_id: str
    parent_node_id: str | None
    depth: int
    tone: str | None
    title: str
    summary: str
    appearing_character_ids: list[str]
    image_path: str | None

    @classmethod
    def from_domain(cls, node: DramaNode) -> DramaNodeResponse:
        return cls(
            id=node.id,
            drama_id=node.drama_id,
            parent_node_id=node.parent_node_id,
            depth=node.depth,
            tone=node.tone,
            title=node.title,
            summary=node.summary,
            appearing_character_ids=list(node.appearing_character_ids),
            image_path=node.image_path,
        )


class CollectedSceneResponse(BaseModel):
    """One picture the player walked past — safe to show whole (BD9)."""

    node_id: str
    title: str
    depth: int
    tone: str | None
    image_path: str

    @classmethod
    def from_domain(cls, scene: CollectedScene) -> CollectedSceneResponse:
        return cls(
            node_id=scene.node_id,
            title=scene.title,
            depth=scene.depth,
            tone=scene.tone,
            image_path=scene.image_path,
        )


class DramaSceneGalleryResponse(BaseModel):
    """劇場圖集 (BD9) — what one player collected out of a painted tree.

    The un-walked pictures are represented by ``locked_count`` alone. That
    is the whole anti-spoiler boundary: this model has no field a title,
    a summary or an image path for a node the player has not reached could
    ride out on, so no future view can leak one by rendering too much.
    """

    collected: list[CollectedSceneResponse]
    locked_count: int
    total_with_images: int
    """Nodes of this drama that currently have a picture — walked or not.

    It grows as lazy generation paints deeper layers, so the same drama
    reports a larger count between visits; that is the tree filling in, not a
    miscount.

    **Not a progress denominator.** It once backed a 「已收集 X/Y」 reading in
    the gallery; that reading is retired, because the prefetch paints branches
    nobody walks into, so this count runs *ahead* of the player instead of
    tracking the walk — the ratio it produced was low and flat and said
    nothing about how much of the drama was left. Completion is measured
    client-side against the whole tree's node count instead. The field's
    meaning and wire format are unchanged."""

    @classmethod
    def from_domain(
        cls, gallery: DramaSceneGallery,
    ) -> DramaSceneGalleryResponse:
        return cls(
            collected=[
                CollectedSceneResponse.from_domain(scene)
                for scene in gallery.collected
            ],
            locked_count=gallery.locked_count,
            total_with_images=gallery.total_with_images,
        )


class ExchangeResponse(BaseModel):
    player_input: str
    response: str

    @classmethod
    def from_domain(cls, ex: Exchange) -> ExchangeResponse:
        return cls(player_input=ex.player_input, response=ex.response)


class DramaSessionTurnResponse(BaseModel):
    node_id: str
    narration: str
    player_input: str
    chosen_tone: str | None
    exchanges: list[ExchangeResponse] = []

    @classmethod
    def from_domain(
        cls, turn: DramaSessionTurn,
    ) -> DramaSessionTurnResponse:
        return cls(
            node_id=turn.node_id,
            narration=turn.narration,
            player_input=turn.player_input,
            chosen_tone=turn.chosen_tone,
            exchanges=[
                ExchangeResponse.from_domain(ex) for ex in turn.exchanges
            ],
        )


class DramaSessionResponse(BaseModel):
    id: str
    drama_id: str
    current_node_id: str
    status: str
    turns: list[DramaSessionTurnResponse]
    created_at: datetime
    updated_at: datetime

    @classmethod
    def from_domain(cls, session: DramaSession) -> DramaSessionResponse:
        return cls(
            id=session.id,
            drama_id=session.drama_id,
            current_node_id=session.current_node_id,
            status=session.status,
            turns=[
                DramaSessionTurnResponse.from_domain(t)
                for t in session.turns
            ],
            created_at=session.created_at,
            updated_at=session.updated_at,
        )


class BranchingDramaResponse(BaseModel):
    id: str
    character_ids: list[str]
    prompt: str
    title: str
    total_segments: int
    status: str
    error_message: str | None = None
    error_code: str | None = None
    """Machine-readable failure reason for a ``failed`` drama (U4b).

    Creation is 202 + poll, so this is the only channel by which a
    player-actionable refusal (``insufficient_credits``) reaches the
    client. ``None`` for ordinary crashes."""
    operator_position: str = DEFAULT_OPERATOR_POSITION
    """Echoed back so the detail view can show the framing the tree was
    planned against, and so BD7's adapt flow can pre-fill it."""
    operator_note: str | None = None
    visual_style: str | None = None
    """The look this drama's scene images are drawn in (BD10), echoed so the
    detail view can show it. ``None`` only for a drama created before the
    slot existed — those still draw off their first character, and saying
    ``anime`` here would name a look that is not what they render."""
    expected_node_count: int
    initial_node_target: int
    """Nodes pre-generated synchronously at creation (root + prefetch
    layers) — the correct denominator for an early progress bar. See
    ``BranchingDrama.initial_node_target()``. ``expected_node_count`` above
    is the full-tree total and stays for the segment-count warning copy."""
    generated_node_count: int = 0
    first_scene_image_path: str | None = None
    first_scene_node_id: str | None = None
    """Root node id, so the detail view's title image can name the node it
    wants redrawn (BD6). ``None`` only while the root does not exist yet —
    the picture may legitimately be missing while the node is not, which is
    exactly the case the redraw button exists to repair."""
    warning: str | None = None
    created_at: datetime
    updated_at: datetime

    @classmethod
    def from_domain(
        cls,
        drama: BranchingDrama,
        *,
        generated_node_count: int = 0,
        first_scene_image_path: str | None = None,
        first_scene_node_id: str | None = None,
    ) -> BranchingDramaResponse:
        warning = None
        if drama.total_segments >= SEGMENTS_WARNING_THRESHOLD:
            count = drama.expected_node_count()
            warning = (
                f"段落數 {drama.total_segments} 將產生 {count} 個節點，"
                f"生成時間可能較長。"
            )
        return cls(
            id=drama.id,
            character_ids=list(drama.character_ids),
            prompt=drama.prompt,
            title=drama.title,
            total_segments=drama.total_segments,
            status=drama.status,
            error_message=drama.error_message,
            error_code=drama.error_code,
            operator_position=drama.operator_position,
            operator_note=drama.operator_note,
            visual_style=drama.visual_style,
            expected_node_count=drama.expected_node_count(),
            initial_node_target=drama.initial_node_target(),
            generated_node_count=generated_node_count,
            first_scene_image_path=first_scene_image_path,
            first_scene_node_id=first_scene_node_id,
            warning=warning,
            created_at=drama.created_at,
            updated_at=drama.updated_at,
        )


class BranchingDramaSummaryResponse(BaseModel):
    id: str
    character_ids: list[str]
    title: str
    total_segments: int
    status: str
    error_message: str | None = None
    error_code: str | None = None
    created_at: datetime
    updated_at: datetime

    @classmethod
    def from_domain(
        cls, drama: BranchingDrama,
    ) -> BranchingDramaSummaryResponse:
        return cls(
            id=drama.id,
            character_ids=list(drama.character_ids),
            title=drama.title,
            total_segments=drama.total_segments,
            status=drama.status,
            error_message=drama.error_message,
            error_code=drama.error_code,
            created_at=drama.created_at,
            updated_at=drama.updated_at,
        )


class InteractSessionResponse(BaseModel):
    """Returned after a player interacts within the current beat."""

    session: DramaSessionResponse
    response: str
    advance_hint: str | None = None


class AdvanceSessionResponse(BaseModel):
    """Returned after a player advances to the next beat."""

    session: DramaSessionResponse
    current_node: DramaNodeResponse
    is_ending: bool
